src/sensors/sonar.py

```diff
@@ -64,58 +75,81 @@ def draw(self):
 
 
 class Sonar(object):
-    def __init__(self, sensor_map, min_range, max_range, cone_angle):
+    """
+    TODO:
+    - add gaussian noise / variance
+    - use OpenGL to draw the rays
+    - Use intensity to simulate min/max range.
+    """
+
+    def __init__(self, sensor_map, min_range, max_range, cone_angle, beams=30):
         self.min_range = min_range
         self.max_range = max_range
         self.cone_angle = cone_angle
         self.sensor_map = sensor_map
+        self.beams = beams
         self.current_range = -1.0
 
     def update_sonar(self, x, y, theta):
-        """Returns the distance to the nearest obstacle for a sensor at position (x, y) and at angle theta."""
-        # start at max range
+        batch = pyglet.graphics.Batch()
+        """Returns the distance to the nearest obstacle for a sensor at
+        position (x, y) and at angle theta."""
+        sensor_pos_map = [int(x / self.sensor_map.resolution),
+                          int(y / self.sensor_map.resolution)]
         self.current_range = self.max_range
-        # create a bundle of rays to replicate a sonar beam
-        sweep = np.arange(-self.cone_angle / 2.0, self.cone_angle / 2.0, SONAR_BEAM_STEP)
-        # cast each ray until it hits an obstacle or the end of the map
-        # print (self.sensor_map.grid)
-        for angle in sweep:
-            distance = 1
-            while distance <= (self.max_range / self.sensor_map.resolution):
-
-                xmap = x / self.sensor_map.resolution
-                xmap += distance * cos(angle + theta)
-                xmap = int(xmap)
-
-                ymap = y / self.sensor_map.resolution
-                ymap += distance * sin(angle + theta)
-                ymap = int(ymap)
-
-                if ymap > self.sensor_map.height - 1 or xmap > self.sensor_map.width - 1:
-                    break
 
-                if ymap < 1 or xmap < 1:
+        step = self.cone_angle / self.beams
+        sweep = np.arange(-self.cone_angle / 2, self.cone_angle / 2, step)
+        sensor_rays = []
+        ray_num = 0
+        #print("sweep")
+        #print(f"sensor position: ({x}, {y})")
+        for angle in sweep:  # Emit rays in a fan
+            x1 = x  # x1 represents the previous x value before collision
+            y1 = y
+            ray = Ray(x, y, SONAR_INTENSITY, angle + theta)
+            while ray.intensity > SONAR_MIN_INTENSITY and ray.bounces < 15:
+                # Continue until ray is no longer detectable or has bounced
+                # too many times
+                xmap = int(ray.x / self.sensor_map.resolution)
+                ymap = int(ray.y / self.sensor_map.resolution)
+                # Check if ray is out of bounds, it should bounce.
+                if xmap < 0 or xmap >= self.sensor_map.width or ymap < 0 or ymap >= self.sensor_map.height:
+                    print(f"Ray {ray_num} (out of bounds) bounced at: ({x1},{y1})!")

```

Comment:
Can we put some kind of if statement around this so it only prints when in debug or developer mode (or similar).  It will be distracting for a child if they happen to be looking at the terminal to see this printed out.